File: articles/mutation-variation-sources-of-novelty/c/mutation_variation_kernel.c
/*
 * Compact mutation and variation kernel in C.
 */

#include <math.h>
#include <stdio.h>

double expected_mutations(double n_genomes, double target_length, double mu) {
    return n_genomes * target_length * mu;
}

double jukes_cantor(double d) {
    if (d >= 0.75) {
        return NAN;
    }

    return -(3.0 / 4.0) * log(1.0 - (4.0 / 3.0) * d);
}

double mutation_selection_balance(double mu, double s) {
    return sqrt(mu / s);
}

int main(void) {
    double p = 0.6;
    double q = 1.0 - p;

    printf("expected_mutations=%.6f\n", expected_mutations(500.0, 1.2e8, 1e-8));
    printf("AA=%.6f\n", p * p);
    printf("Aa=%.6f\n", 2.0 * p * q);
    printf("aa=%.6f\n", q * q);
    printf("jukes_cantor=%.6f\n", jukes_cantor(0.15));
    printf("mutation_selection_balance_q=%.6f\n", mutation_selection_balance(1e-5, 0.01));
    printf("pi_site=%.6f\n", 2.0 * p * (1.0 - p));

    return 0;
}

File: articles/mutation-variation-sources-of-novelty/fortran/mutation_variation_kernel.f90
! Compact mutation and variation numerical kernel in Fortran.

program mutation_variation_kernel
  implicit none

  real :: mu
  real :: L
  real :: n_genomes
  real :: lambda
  real :: p
  real :: q
  real :: d
  real :: d_jc
  real :: sel
  real :: q_star
  real :: pi_site

  mu = 1.0e-8
  L = 1.2e8
  n_genomes = 500.0

  lambda = n_genomes * L * mu

  p = 0.6
  q = 1.0 - p

  d = 0.15
  d_jc = -(3.0 / 4.0) * log(1.0 - (4.0 / 3.0) * d)

  sel = 0.01
  q_star = sqrt(1.0e-5 / sel)

  pi_site = 2.0 * p * (1.0 - p)

  print *, "Expected mutations:", lambda
  print *, "Hardy-Weinberg AA:", p ** 2
  print *, "Hardy-Weinberg Aa:", 2.0 * p * q
  print *, "Hardy-Weinberg aa:", q ** 2
  print *, "Jukes-Cantor distance:", d_jc
  print *, "Mutation-selection balance q:", q_star
  print *, "Pi at p=0.6:", pi_site
end program mutation_variation_kernel
